File: ai_scrapper/verifier.py
"""
verifier.py — Email & phone credibility verification for scraped business data.

Email verification (in order, cheapest first — stops early on a clear fail):
  1. Syntax check (regex)
  2. Disposable/fake-pattern check (test@, example.com, noreply@, etc.)
  3. MX record lookup (does the domain even accept mail?)
  4. SMTP RCPT TO probe (does the mailbox itself exist?)

Phone verification (via the `phonenumbers` library, Google's libphonenumber port):
  1. Parse with a default region hint
  2. is_valid_number() — correct length/structure for that region
  3. number_type() — mobile / fixed line / voip / etc.
  4. region/carrier description

Design notes:
- SMTP checks are unreliable by nature: many providers (Gmail, Outlook, Microsoft 365)
  greylist or block probing connections, or accept-all to avoid leaking which addresses
  are real. Results include an explicit "unknown" status — never silently coerced to
  True or False.
- SMTP checks are rate-limited separately from page-crawl delays to avoid getting the
  scraping IP flagged by mail servers.
- MX/SMTP results are cached per-domain within a run, since many emails on one company
  page usually share a domain — no need to repeat the same MX lookup or SMTP handshake.
"""

import re
import secrets
import smtplib
import socket
import time
from collections.abc import Callable
import dns.resolver
import phonenumbers
from phonenumbers import number_type, PhoneNumberType, geocoder, carrier as ph_carrier

# ============================================================
# Config
# ============================================================
DEFAULT_REGION = "FR"          # used to parse phone numbers with no country code
SMTP_TIMEOUT = 8                # seconds per SMTP connection attempt
SMTP_DELAY_SECONDS = 1.5        # politeness delay between SMTP probes (separate from page delay)
SMTP_FROM_ADDRESS = "verify@example.com"   # used in MAIL FROM during the probe
SMTP_HELO_DOMAIN = "example.com"

DISPOSABLE_OR_FAKE_PATTERNS = [
    r"^test@", r"^example@", r"^demo@", r"^sample@", r"^noreply@", r"^no-reply@",
    r"^donotreply@", r"^admin@example\.", r"^user@", r"^foo@", r"^bar@",
    r"@example\.(com|org|net)$", r"@test\.(com|local)$", r"@domain\.(com|tld)$",
    r"@yourcompany\.", r"@yourdomain\.", r"@company\.com$",
]

EMAIL_REGEX = re.compile(r"^[\w\.\-+]+@[\w\.\-]+\.\w+$")

FREE_EMAIL_DOMAINS = {
    "gmail.com", "googlemail.com", "yahoo.com", "yahoo.fr", "outlook.com", "live.com", "hotmail.com",
    "icloud.com", "me.com", "proton.me", "protonmail.com", "aol.com", "yandex.com", "gmx.com",
    "mail.com", "zoho.com", "tuta.io", "tutanota.com",
}

DISPOSABLE_DOMAINS = {
    "mailinator.com", "yopmail.com", "tempmail.com", "10minutemail.com", "guerrillamail.com",
    "throwaway.email", "fakeinbox.com", "trashmail.com", "maildrop.cc", "getnada.com",
    "temp-mail.org", "sharklasers.com", "disposable.com",
}

ROLE_BASED_LOCAL_PARTS = {
    "admin", "administration", "contact", "info", "support", "sales", "vente", "secretariat",
    "secretaire", "direction", "dg", "rh", "hr", "compta", "finance", "accueil", "standard",
    "noreply", "no-reply", "donotreply",
}

_mx_cache = {}      # domain -> (has_mx: bool, mx_host: str|None)
_smtp_cache = {}     # email -> result dict (avoid re-probing same address twice in a run)
_txt_cache = {}      # domain -> list[str]


# ============================================================
# Email verification
# ============================================================
def _email_domain(email: str) -> str:
    return email.split("@")[-1].lower()


def _looks_fake(email: str) -> bool:
    email_lower = email.lower()
    domain = _email_domain(email)
    return any(re.search(p, email_lower) for p in DISPOSABLE_OR_FAKE_PATTERNS) or domain in DISPOSABLE_DOMAINS


def _is_free_provider(domain: str) -> bool:
    return domain in FREE_EMAIL_DOMAINS


def _is_role_based(email: str) -> bool:
    local_part = email.split("@")[0].lower()
    normalized = re.sub(r"[^a-z0-9]", "", local_part)
    return normalized in ROLE_BASED_LOCAL_PARTS or any(re.search(rf"(^|[^a-z0-9]){re.escape(part)}([^a-z0-9]|$)", local_part) for part in ROLE_BASED_LOCAL_PARTS)


def _get_txt_records(domain: str) -> list[str]:
    if domain in _txt_cache:
        return _txt_cache[domain]
    try:
        answers = dns.resolver.resolve(domain, "TXT", lifetime=8)
        _txt_cache[domain] = [str(answer).strip('"') for answer in answers]
    except Exception:
        _txt_cache[domain] = []
    return _txt_cache[domain]


def _has_spf(domain: str) -> bool:
    return any(record.startswith("v=spf1") for record in _get_txt_records(domain))


def _has_dmarc(domain: str) -> bool:
    try:
        answers = dns.resolver.resolve(f"_dmarc.{domain}", "TXT", lifetime=8)
        return any(str(answer).strip('"').lower().startswith("v=dmarc1") for answer in answers)
    except Exception:
        return False


def _get_mx_host(domain: str):
    """Look up MX record for a domain. Cached. Returns (has_mx, mx_host)."""
    if domain in _mx_cache:
        return _mx_cache[domain]
    try:
        answers = dns.resolver.resolve(domain, "MX", lifetime=8)
        mx_hosts = sorted(answers, key=lambda r: r.preference)
        mx_host = str(mx_hosts[0].exchange).rstrip(".")
        _mx_cache[domain] = (True, mx_host)
    except Exception:
        _mx_cache[domain] = (False, None)
    return _mx_cache[domain]


def _smtp_probe(email: str, mx_host: str):
    """
    Connect to the mail server and ask if the mailbox exists, without sending mail.
    Returns one of: "valid", "invalid", "unknown" (+ a human-readable reason).

    This is inherently unreliable — many providers won't give a straight answer.
    "unknown" is a legitimate, expected outcome, not an error state.
    """
    if email in _smtp_cache:
        return _smtp_cache[email]

    result = {"smtp_status": "unknown", "smtp_detail": ""}
    server = None
    try:
        time.sleep(SMTP_DELAY_SECONDS)  # politeness / avoid being flagged
        server = smtplib.SMTP(timeout=SMTP_TIMEOUT)
        server.connect(mx_host)
        server.helo(SMTP_HELO_DOMAIN)
        server.mail(SMTP_FROM_ADDRESS)
        code, message = server.rcpt(email)

        if code == 250:
            result = {"smtp_status": "valid", "smtp_detail": "mailbox accepted"}
        elif code in (550, 551, 553):
            result = {"smtp_status": "invalid", "smtp_detail": f"rejected (code {code})"}
        else:
            result = {"smtp_status": "unknown", "smtp_detail": f"ambiguous response (code {code})"}

    except (smtplib.SMTPServerDisconnected, smtplib.SMTPConnectError, socket.timeout, OSError) as e:
        result = {"smtp_status": "unknown", "smtp_detail": f"server blocked probe or unreachable: {e}"}
    except smtplib.SMTPRecipientsRefused:
        result = {"smtp_status": "invalid", "smtp_detail": "recipient refused"}
    except Exception as e:
        result = {"smtp_status": "unknown", "smtp_detail": f"probe failed: {e}"}
    finally:
        if server:
            try:
                server.quit()
            except Exception:
                pass

    _smtp_cache[email] = result
    return result


def _smtp_probe_random(mx_host: str, domain: str) -> dict:
    random_email = f"kilo_verify_{secrets.token_hex(4)}@{domain}"
    return _smtp_probe(random_email, mx_host)


def _score_email(checks: dict) -> tuple[int, str, str, str]:
    if checks["disposable_or_fake"]:
        return 0, "invalid", "high", "disposable or fake-pattern domain"
    if not checks["syntax_valid"]:
        return 0, "invalid", "high", "invalid email syntax"
    if not checks["mx_record"]:
        return 20, "invalid", "high", "domain has no MX record"

    score = 40
    if checks["spf"] == "present":
        score += 8
    elif checks["spf"] == "missing":
        score -= 5
    if checks["dmarc"] == "present":
        score += 7
    elif checks["dmarc"] == "missing":
        score -= 3
    if checks["free_provider"]:
        score -= 5
    if checks["role_based"]:
        score -= 3

    smtp_status = checks["smtp_status"]
    catch_all = checks["catch_all"]
    if smtp_status == "valid":
        score += 35
        if catch_all:
            score -= 10
            deliverability = "risky"
            reason = "SMTP accepted, but domain appears catch-all"
        else:
            deliverability = "confirmed"
            reason = "SMTP accepted the mailbox"
    elif smtp_status == "invalid":
        return 0, "invalid", "high", "SMTP rejected the mailbox"
    elif smtp_status == "unknown":
        score += 18
        deliverability = "uncertain"
        reason = "SMTP probe was inconclusive"
    else:
        deliverability = "likely" if checks["spf"] == "present" and checks["dmarc"] == "present" else "uncertain"
        reason = "Syntax + MX verified; SMTP not run"

    score = max(0, min(100, score))
    if score >= 80:
        credibility = "high"
    elif score >= 65:
        credibility = "medium"
    elif score >= 40:
        credibility = "low"
    else:
        credibility = "rejected"

    risk = "low" if score >= 80 else "medium" if score >= 55 else "high"
    return score, deliverability, risk, reason


def credibility_from_score(score: int) -> str:
    if score >= 80:
        return "high"
    if score >= 65:
        return "medium"
    if score >= 40:
        return "low"
    return "rejected"


def verify_email(email: str, do_smtp: bool = True, emit: Callable | None = None, check_catch_all: bool = True) -> dict:
    """
    Returns a dict:
    {
        "email": str,
        "syntax_valid": bool,
        "is_disposable_or_fake_pattern": bool,
        "has_mx_record": bool | None,
        "mx_host": str | None,
        "smtp_status": "valid" | "invalid" | "unknown" | "skipped",
        "smtp_detail": str,
        "credibility": "high" | "medium" | "low" | "rejected",
        "score": int,
        "deliverability": "confirmed" | "likely" | "uncertain" | "risky" | "invalid",
        "risk": "low" | "medium" | "high",
        "reason": str,
        "checks": dict,
    }
    """
    email = email.strip()
    out = {
        "email": email,
        "syntax_valid": False,
        "is_disposable_or_fake_pattern": False,
        "has_mx_record": None,
        "mx_host": None,
        "smtp_status": "skipped",
        "smtp_detail": "",
        "credibility": "rejected",
        "score": 0,
        "deliverability": "uncertain",
        "risk": "high",
        "reason": "",
        "checks": {},
    }

    def notify(kind: str, payload: dict | None = None):
        if emit:
            emit(kind, {"email": email, **(payload or {})})

    # 1. Syntax
    out["syntax_valid"] = bool(EMAIL_REGEX.match(email))
    if not out["syntax_valid"]:
        notify("email_syntax_failed")
        out["credibility"] = "rejected"
        return out
    notify("email_syntax_valid")

    # 2. Fake/disposable pattern check
    out["is_disposable_or_fake_pattern"] = _looks_fake(email)
    if out["is_disposable_or_fake_pattern"]:
        notify("email_fake_pattern")
        out["credibility"] = "rejected"
        return out

    # 3. MX record
    domain = _email_domain(email)
    has_mx, mx_host = _get_mx_host(domain)
    spf_present = _has_spf(domain) if has_mx else False
    dmarc_present = _has_dmarc(domain) if has_mx else False
    free_provider = _is_free_provider(domain)
    role_based = _is_role_based(email)
    out["has_mx_record"] = has_mx
    out["mx_host"] = mx_host
    notify("email_mx_checked", {"domain": domain, "has_mx_record": has_mx, "mx_host": mx_host})
    notify("email_dns_checked", {"domain": domain, "spf": spf_present, "dmarc": dmarc_present, "free_provider": free_provider, "role_based": role_based})

    checks = {
        "syntax_valid": out["syntax_valid"],
        "disposable_or_fake": out["is_disposable_or_fake_pattern"],
        "free_provider": free_provider,
        "role_based": role_based,
        "mx_record": bool(has_mx),
        "mx_host": mx_host,
        "spf": "present" if spf_present else "missing" if has_mx else "unchecked",
        "dmarc": "present" if dmarc_present else "missing" if has_mx else "unchecked",
        "smtp_status": "skipped",
        "smtp_detail": "",
        "catch_all": None,
    }

    if not has_mx:
        notify("email_no_mx")
        score, deliverability, risk, reason = _score_email(checks)
        out.update({"credibility": "low", "score": score, "deliverability": deliverability, "risk": risk, "reason": reason, "checks": checks})
        return out

    # 4. SMTP probe (optional — slower, sometimes blocked)
    if do_smtp:
        notify("email_smtp_probe_start")
        smtp_result = _smtp_probe(email, mx_host)
        checks["smtp_status"] = smtp_result["smtp_status"]
        checks["smtp_detail"] = smtp_result["smtp_detail"]
        out["smtp_status"] = smtp_result["smtp_status"]
        out["smtp_detail"] = smtp_result["smtp_detail"]
        notify("email_smtp_probe_done", {"smtp_status": smtp_result["smtp_status"], "smtp_detail": smtp_result["smtp_detail"]})

        if smtp_result["smtp_status"] == "valid" and check_catch_all:
            notify("email_catch_all_probe_start")
            catch_all_result = _smtp_probe_random(mx_host, domain)
            checks["catch_all"] = catch_all_result["smtp_status"] == "valid"
            notify("email_catch_all_probe_done", {"catch_all": checks["catch_all"], "probe_status": catch_all_result["smtp_status"], "probe_detail": catch_all_result["smtp_detail"]})
    else:
        notify("email_smtp_skipped")

    score, deliverability, risk, reason = _score_email(checks)
    out.update({"credibility": credibility_from_score(score), "score": score, "deliverability": deliverability, "risk": risk, "reason": reason, "checks": checks})
    notify("email_scored", {"score": score, "credibility": out["credibility"], "deliverability": deliverability, "risk": risk, "reason": reason})

    return out


# ============================================================
# Phone verification
# ============================================================
_TYPE_NAMES = {
    PhoneNumberType.FIXED_LINE: "fixed_line",
    PhoneNumberType.MOBILE: "mobile",
    PhoneNumberType.FIXED_LINE_OR_MOBILE: "fixed_line_or_mobile",
    PhoneNumberType.TOLL_FREE: "toll_free",
    PhoneNumberType.PREMIUM_RATE: "premium_rate",
    PhoneNumberType.SHARED_COST: "shared_cost",
    PhoneNumberType.VOIP: "voip",
    PhoneNumberType.PERSONAL_NUMBER: "personal_number",
    PhoneNumberType.PAGER: "pager",
    PhoneNumberType.UAN: "uan",
    PhoneNumberType.VOICEMAIL: "voicemail",
    PhoneNumberType.UNKNOWN: "unknown",
}


def verify_phone(raw_number: str, default_region: str = DEFAULT_REGION, emit: Callable | None = None) -> dict:
    """
    Returns a dict:
    {
        "raw": str,
        "parsed_e164": str | None,
        "is_valid": bool,
        "is_possible": bool,
        "number_type": str,
        "region": str | None,
        "carrier": str | None,
        "credibility": "high" | "medium" | "rejected",
    }
    """
    out = {
        "raw": raw_number,
        "parsed_e164": None,
        "is_valid": False,
        "is_possible": False,
        "number_type": "unknown",
        "region": None,
        "carrier": None,
        "credibility": "rejected",
    }

    def notify(kind: str, payload: dict | None = None):
        if emit:
            emit(kind, {"raw": raw_number, **(payload or {})})

    try:
        parsed = phonenumbers.parse(raw_number, default_region)
    except phonenumbers.NumberParseException as e:
        notify("phone_parse_failed", {"error": str(e)})
        out["credibility"] = "rejected"
        return out

    out["is_possible"] = phonenumbers.is_possible_number(parsed)
    out["is_valid"] = phonenumbers.is_valid_number(parsed)
    notify("phone_parsed", {"is_possible": out["is_possible"], "is_valid": out["is_valid"]})

    if not out["is_valid"]:
        out["credibility"] = "rejected" if not out["is_possible"] else "low"
        notify("phone_invalid")
        return out

    out["parsed_e164"] = phonenumbers.format_number(parsed, phonenumbers.PhoneNumberFormat.E164)
    out["number_type"] = _TYPE_NAMES.get(number_type(parsed), "unknown")
    out["region"] = geocoder.description_for_number(parsed, "en") or None

    try:
        out["carrier"] = ph_carrier.name_for_number(parsed, "en") or None
    except Exception:
        out["carrier"] = None

    notify("phone_verified", {"parsed_e164": out["parsed_e164"], "number_type": out["number_type"], "region": out["region"], "carrier": out["carrier"]})

    # Valid + real structure = high confidence. (No live reachability check, that's a paid-API tier.)
    out["credibility"] = "high"
    return out


# ============================================================
# Batch helpers — used by the crawler
# ============================================================
def verify_emails(emails: list, do_smtp: bool = True, emit: Callable | None = None, check_catch_all: bool = True) -> list:
    return [verify_email(e, do_smtp=do_smtp, emit=emit, check_catch_all=check_catch_all) for e in emails]


def verify_phones(phones: list, default_region: str = DEFAULT_REGION, emit: Callable | None = None) -> list:
    return [verify_phone(p, default_region=default_region, emit=emit) for p in phones]


if __name__ == "__main__":
    # Quick manual test
    test_emails = ["contact@kompass.com", "test@example.com", "fake@nonexistentdomain12345.com"]
    test_phones = ["+33 1 42 68 53 00", "0142685300", "not a phone"]

    print("--- Email tests ---")
    for e in test_emails:
        print(verify_email(e, do_smtp=True))

    print("\n--- Phone tests ---")
    for p in test_phones:
        print(verify_phone(p))

File: apps/web/src/lib/scraper/verifier.ts
import dns from 'node:dns/promises';
import net from 'node:net';
import { parsePhoneNumberFromString } from 'libphonenumber-js';

/**
 * Email & phone credibility verification for scraped business contact data.
 * Ported from the Python `ai_scrapper/verifier.py` reference implementation.
 *
 * Email verification (cheapest first, stops early on a clear fail):
 *   1. Syntax check (regex)
 *   2. Disposable/fake-pattern check (test@, example.com, noreply@, etc.)
 *   3. MX record lookup (does the domain even accept mail?)
 *   4. SPF/DMARC presence
 *   5. (optional) SMTP RCPT TO probe — best-effort only. Many mail providers
 *      block or greylist probing connections, and outbound port 25 is often
 *      blocked on serverless platforms, so "unknown" is a legitimate outcome.
 */

const SMTP_TIMEOUT_MS = 5000;
const SMTP_FROM_ADDRESS = 'verify@example.com';
const SMTP_HELO_DOMAIN = 'example.com';

const DISPOSABLE_OR_FAKE_PATTERNS = [
  /^test@/i, /^example@/i, /^demo@/i, /^sample@/i, /^noreply@/i, /^no-reply@/i,
  /^donotreply@/i, /^admin@example\./i, /^user@/i, /^foo@/i, /^bar@/i,
  /@example\.(com|org|net)$/i, /@test\.(com|local)$/i, /@domain\.(com|tld)$/i,
  /@yourcompany\./i, /@yourdomain\./i, /@company\.com$/i,
];

const EMAIL_REGEX = /^[\w.\-+]+@[\w.-]+\.\w+$/;

const FREE_EMAIL_DOMAINS = new Set([
  'gmail.com', 'googlemail.com', 'yahoo.com', 'yahoo.fr', 'outlook.com', 'live.com', 'hotmail.com',
  'icloud.com', 'me.com', 'proton.me', 'protonmail.com', 'aol.com', 'yandex.com', 'gmx.com',
  'mail.com', 'zoho.com', 'tuta.io', 'tutanota.com',
]);

const DISPOSABLE_DOMAINS = new Set([
  'mailinator.com', 'yopmail.com', 'tempmail.com', '10minutemail.com', 'guerrillamail.com',
  'throwaway.email', 'fakeinbox.com', 'trashmail.com', 'maildrop.cc', 'getnada.com',
  'temp-mail.org', 'sharklasers.com', 'disposable.com',
]);

const ROLE_BASED_LOCAL_PARTS = new Set([
  'admin', 'administration', 'contact', 'info', 'support', 'sales', 'vente', 'secretariat',
  'secretaire', 'direction', 'dg', 'rh', 'hr', 'compta', 'finance', 'accueil', 'standard',
  'noreply', 'no-reply', 'donotreply',
]);

const mxCache = new Map<string, { hasMx: boolean; mxHost: string | null }>();
const txtCache = new Map<string, string[]>();

export interface EmailVerificationResult {
  email: string;
  syntaxValid: boolean;
  isDisposableOrFakePattern: boolean;
  hasMxRecord: boolean | null;
  mxHost: string | null;
  smtpStatus: 'valid' | 'invalid' | 'unknown' | 'skipped';
  smtpDetail: string;
  credibility: 'high' | 'medium' | 'low' | 'rejected';
  score: number;
  deliverability: 'confirmed' | 'likely' | 'uncertain' | 'risky' | 'invalid';
  risk: 'low' | 'medium' | 'high';
  reason: string;
  checks: Record<string, unknown>;
}

export interface PhoneVerificationResult {
  raw: string;
  parsedE164: string | null;
  isValid: boolean;
  isPossible: boolean;
  numberType: string;
  region: string | null;
  carrier: string | null;
  credibility: 'high' | 'low' | 'rejected';
}

function emailDomain(email: string): string {
  return email.split('@').pop()!.toLowerCase();
}

function looksFake(email: string): boolean {
  const lower = email.toLowerCase();
  const domain = emailDomain(email);
  return DISPOSABLE_OR_FAKE_PATTERNS.some((p) => p.test(lower)) || DISPOSABLE_DOMAINS.has(domain);
}

function isFreeProvider(domain: string): boolean {
  return FREE_EMAIL_DOMAINS.has(domain);
}

function isRoleBased(email: string): boolean {
  const localPart = email.split('@')[0]!.toLowerCase();
  const normalized = localPart.replace(/[^a-z0-9]/g, '');
  return ROLE_BASED_LOCAL_PARTS.has(normalized);
}

async function getTxtRecords(domain: string): Promise<string[]> {
  if (txtCache.has(domain)) return txtCache.get(domain)!;
  try {
    const records = await dns.resolveTxt(domain);
    const flat = records.map((chunks) => chunks.join(''));
    txtCache.set(domain, flat);
    return flat;
  } catch {
    txtCache.set(domain, []);
    return [];
  }
}

async function hasSpf(domain: string): Promise<boolean> {
  const records = await getTxtRecords(domain);
  return records.some((r) => r.startsWith('v=spf1'));
}

async function hasDmarc(domain: string): Promise<boolean> {
  try {
    const records = await dns.resolveTxt(`_dmarc.${domain}`);
    return records.some((chunks) => chunks.join('').toLowerCase().startsWith('v=dmarc1'));
  } catch {
    return false;
  }
}

async function getMxHost(domain: string): Promise<{ hasMx: boolean; mxHost: string | null }> {
  if (mxCache.has(domain)) return mxCache.get(domain)!;
  try {
    const records = await dns.resolveMx(domain);
    if (records.length === 0) throw new Error('no mx');
    records.sort((a, b) => a.priority - b.priority);
    const result = { hasMx: true, mxHost: records[0]!.exchange };
    mxCache.set(domain, result);
    return result;
  } catch {
    const result = { hasMx: false, mxHost: null };
    mxCache.set(domain, result);
    return result;
  }
}

/**
 * Best-effort SMTP RCPT TO probe. Many networks (including serverless
 * platforms) block outbound port 25 — a timeout/refusal is treated as
 * "unknown", never coerced to true/false.
 */
function smtpProbe(email: string, mxHost: string): Promise<{ smtpStatus: 'valid' | 'invalid' | 'unknown'; smtpDetail: string }> {
  return new Promise((resolve) => {
    let settled = false;
    let stage = 0; // 0=connect,1=helo,2=mail,3=rcpt
    const socket = net.createConnection({ host: mxHost, port: 25, timeout: SMTP_TIMEOUT_MS });

    const finish = (result: { smtpStatus: 'valid' | 'invalid' | 'unknown'; smtpDetail: string }) => {
      if (settled) return;
      settled = true;
      try {
        socket.write('QUIT\r\n');
      } catch {
        /* ignore */
      }
      socket.destroy();
      resolve(result);
    };

    socket.setTimeout(SMTP_TIMEOUT_MS, () => {
      finish({ smtpStatus: 'unknown', smtpDetail: 'connection timed out' });
    });

    socket.on('error', (err) => {
      finish({ smtpStatus: 'unknown', smtpDetail: `server blocked probe or unreachable: ${err.message}` });
    });

    socket.on('data', (data) => {
      const line = data.toString();
      const code = parseInt(line.slice(0, 3), 10);

      if (stage === 0 && code === 220) {
        stage = 1;
        socket.write(`HELO ${SMTP_HELO_DOMAIN}\r\n`);
      } else if (stage === 1 && code === 250) {
        stage = 2;
        socket.write(`MAIL FROM:<${SMTP_FROM_ADDRESS}>\r\n`);
      } else if (stage === 2 && code === 250) {
        stage = 3;
        socket.write(`RCPT TO:<${email}>\r\n`);
      } else if (stage === 3) {
        if (code === 250) {
          finish({ smtpStatus: 'valid', smtpDetail: 'mailbox accepted' });
        } else if (code === 550 || code === 551 || code === 553) {
          finish({ smtpStatus: 'invalid', smtpDetail: `rejected (code ${code})` });
        } else {
          finish({ smtpStatus: 'unknown', smtpDetail: `ambiguous response (code ${code})` });
        }
      } else if (code >= 400) {
        finish({ smtpStatus: 'unknown', smtpDetail: `ambiguous response (code ${code})` });
      }
    });
  });
}

function scoreEmail(checks: Record<string, any>): { score: number; deliverability: EmailVerificationResult['deliverability']; risk: EmailVerificationResult['risk']; reason: string } {
  if (checks.disposableOrFake) return { score: 0, deliverability: 'invalid', risk: 'high', reason: 'disposable or fake-pattern domain' };
  if (!checks.syntaxValid) return { score: 0, deliverability: 'invalid', risk: 'high', reason: 'invalid email syntax' };
  if (!checks.mxRecord) return { score: 20, deliverability: 'invalid', risk: 'high', reason: 'domain has no MX record' };

  let score = 40;
  if (checks.spf === 'present') score += 8;
  else if (checks.spf === 'missing') score -= 5;
  if (checks.dmarc === 'present') score += 7;
  else if (checks.dmarc === 'missing') score -= 3;
  if (checks.freeProvider) score -= 5;
  if (checks.roleBased) score -= 3;

  let deliverability: EmailVerificationResult['deliverability'];
  let reason: string;

  if (checks.smtpStatus === 'valid') {
    score += 35;
    if (checks.catchAll) {
      score -= 10;
      deliverability = 'risky';
      reason = 'SMTP accepted, but domain appears catch-all';
    } else {
      deliverability = 'confirmed';
      reason = 'SMTP accepted the mailbox';
    }
  } else if (checks.smtpStatus === 'invalid') {
    return { score: 0, deliverability: 'invalid', risk: 'high', reason: 'SMTP rejected the mailbox' };
  } else if (checks.smtpStatus === 'unknown') {
    score += 18;
    deliverability = 'uncertain';
    reason = 'SMTP probe was inconclusive';
  } else {
    deliverability = checks.spf === 'present' && checks.dmarc === 'present' ? 'likely' : 'uncertain';
    reason = 'Syntax + MX verified; SMTP not run';
  }

  score = Math.max(0, Math.min(100, score));
  const risk: EmailVerificationResult['risk'] = score >= 80 ? 'low' : score >= 55 ? 'medium' : 'high';
  return { score, deliverability, risk, reason };
}

export function credibilityFromScore(score: number): EmailVerificationResult['credibility'] {
  if (score >= 80) return 'high';
  if (score >= 65) return 'medium';
  if (score >= 40) return 'low';
  return 'rejected';
}

export async function verifyEmail(rawEmail: string, doSmtp = true, checkCatchAll = true): Promise<EmailVerificationResult> {
  const email = rawEmail.trim();
  const out: EmailVerificationResult = {
    email,
    syntaxValid: false,
    isDisposableOrFakePattern: false,
    hasMxRecord: null,
    mxHost: null,
    smtpStatus: 'skipped',
    smtpDetail: '',
    credibility: 'rejected',
    score: 0,
    deliverability: 'uncertain',
    risk: 'high',
    reason: '',
    checks: {},
  };

  out.syntaxValid = EMAIL_REGEX.test(email);
  if (!out.syntaxValid) {
    out.credibility = 'rejected';
    out.reason = 'invalid email syntax';
    return out;
  }

  out.isDisposableOrFakePattern = looksFake(email);
  if (out.isDisposableOrFakePattern) {
    out.credibility = 'rejected';
    out.reason = 'disposable or fake-pattern domain';
    return out;
  }

  const domain = emailDomain(email);
  const { hasMx, mxHost } = await getMxHost(domain);
  const spfPresent = hasMx ? await hasSpf(domain) : false;
  const dmarcPresent = hasMx ? await hasDmarc(domain) : false;
  const freeProvider = isFreeProvider(domain);
  const roleBased = isRoleBased(email);

  out.hasMxRecord = hasMx;
  out.mxHost = mxHost;

  const checks: Record<string, any> = {
    syntaxValid: out.syntaxValid,
    disposableOrFake: out.isDisposableOrFakePattern,
    freeProvider,
    roleBased,
    mxRecord: hasMx,
    mxHost,
    spf: hasMx ? (spfPresent ? 'present' : 'missing') : 'unchecked',
    dmarc: hasMx ? (dmarcPresent ? 'present' : 'missing') : 'unchecked',
    smtpStatus: 'skipped',
    smtpDetail: '',
    catchAll: null,
  };

  if (!hasMx) {
    const { score, deliverability, risk, reason } = scoreEmail(checks);
    Object.assign(out, { credibility: 'low', score, deliverability, risk, reason, checks });
    return out;
  }

  if (doSmtp && mxHost) {
    const smtpResult = await smtpProbe(email, mxHost);
    checks.smtpStatus = smtpResult.smtpStatus;
    checks.smtpDetail = smtpResult.smtpDetail;
    out.smtpStatus = smtpResult.smtpStatus;
    out.smtpDetail = smtpResult.smtpDetail;

    if (smtpResult.smtpStatus === 'valid' && checkCatchAll) {
      const randomLocal = `probe_${Math.random().toString(36).slice(2, 10)}`;
      const catchAllResult = await smtpProbe(`${randomLocal}@${domain}`, mxHost);
      checks.catchAll = catchAllResult.smtpStatus === 'valid';
    }
  }

  const { score, deliverability, risk, reason } = scoreEmail(checks);
  Object.assign(out, { credibility: credibilityFromScore(score), score, deliverability, risk, reason, checks });
  return out;
}

export async function verifyEmails(emails: string[], doSmtp = true, checkCatchAll = true): Promise<EmailVerificationResult[]> {
  const results: EmailVerificationResult[] = [];
  for (const email of emails) {
    results.push(await verifyEmail(email, doSmtp, checkCatchAll));
  }
  return results;
}

export function verifyPhone(rawNumber: string, defaultRegion: string = 'FR'): PhoneVerificationResult {
  const out: PhoneVerificationResult = {
    raw: rawNumber,
    parsedE164: null,
    isValid: false,
    isPossible: false,
    numberType: 'unknown',
    region: null,
    carrier: null,
    credibility: 'rejected',
  };

  try {
    const phone = parsePhoneNumberFromString(rawNumber, defaultRegion as any);
    if (!phone) return out;

    out.isPossible = phone.isPossible();
    out.isValid = phone.isValid();

    if (!out.isValid) {
      out.credibility = out.isPossible ? 'low' : 'rejected';
      return out;
    }

    out.parsedE164 = phone.formatInternational().replace(/\s+/g, '');
    out.numberType = phone.getType() || 'unknown';
    out.region = phone.country || null;
    out.credibility = 'high';
    return out;
  } catch {
    return out;
  }
}

export function verifyPhones(phones: string[], defaultRegion = 'FR'): PhoneVerificationResult[] {
  return phones.map((p) => verifyPhone(p, defaultRegion));
}
